ext/random/random.c

```diff
@@ -0,0 +1,1785 @@
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	HashTable *data;
+	zval *tmp;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_HT(data);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* members */
+	tmp = zend_hash_index_find(data, 0);
+	if (!tmp || Z_TYPE_P(tmp) != IS_ARRAY) {
+		zend_throw_exception(NULL, "Incomplete or ill-formed serialization data", 0);
+		RETURN_THROWS();
+	}
+	object_properties_load(&engine->std, Z_ARRVAL_P(tmp));
+
+	/* state */
+	tmp = zend_hash_index_find(data, 1);
+	if (!tmp || Z_TYPE_P(tmp) != IS_ARRAY) {
+		zend_throw_exception(NULL, "Incomplete or ill-formed serialization data", 0);
+		RETURN_THROWS();
+	}
+	if (engine->algo->unserialize(engine->state, Z_ARRVAL_P(tmp)) == FAILURE) {
+		zend_throw_exception(NULL, "Engine unserialize failed", 0);
+		RETURN_THROWS();
+	}
+}
+/* }}} */
+
+/* {{{ DebugInfo */
+PHP_METHOD(Random_Engine_CombinedLCG, __debugInfo)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	zval tmp;
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	if (!engine->std.properties) {
+		rebuild_object_properties(&engine->std);
+	}
+	ZVAL_ARR(return_value, zend_array_dup(engine->std.properties));
+
+	if (engine->algo->serialize) {
+		array_init(&tmp);
+		if (engine->algo->serialize(engine->state, Z_ARRVAL(tmp)) == FAILURE) {
+			zend_throw_exception(NULL, "Engine serialize failed", 0);
+			RETURN_THROWS();
+		}
+		zend_hash_str_add(Z_ARR_P(return_value), "__states", sizeof("__states") - 1, &tmp);
+	}
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Engine_MersenneTwister, __construct)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	php_random_engine_state_mersennetwister *state = engine->state;
+	zend_long seed, mode = MT_RAND_MT19937;
+	bool seed_is_null = true;
+
+	ZEND_PARSE_PARAMETERS_START(0, 2)
+		Z_PARAM_OPTIONAL;
+		Z_PARAM_LONG_OR_NULL(seed, seed_is_null);
+		Z_PARAM_LONG(mode);
+	ZEND_PARSE_PARAMETERS_END();
+
+	switch (mode) {
+		case MT_RAND_PHP:
+			state->mode = MT_RAND_PHP;
+			break;
+		default:
+			state->mode = MT_RAND_MT19937;
+	}
+
+	if (seed_is_null) {
+		/* MT19937 has a very large state, uses CSPRNG for seeding only */
+		if (php_random_bytes_silent(&seed, sizeof(zend_long)) == FAILURE) {
+			zend_throw_exception(spl_ce_RuntimeException, "Random number generate failed", 0);
+			RETURN_THROWS();
+		}
+	}
+
+	engine->algo->seed(engine->state, seed);
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Engine_PCG64, __construct)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	php_random_engine_state_pcg64 *state = engine->state;
+	zend_string *str_seed = NULL;
+	zend_long int_seed = 0;
+	bool seed_is_null = true;
+	int i, j;
+	uint64_t t[2];
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL;
+		Z_PARAM_STR_OR_LONG_OR_NULL(str_seed, int_seed, seed_is_null);
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (seed_is_null) {
+		if (php_random_bytes_silent(&state->inc, sizeof(random_uint128_t)) == FAILURE) {
```

Comment:
```suggestion
		if (php_random_bytes_silent(&state->s, sizeof(random_uint128_t)) == FAILURE) {
```
? See also my email.